Framework/Algorithms/src/Integration.cpp

```diff
@@ -278,19 +279,35 @@ void Integration::exec() {
         if (rebinned_input) {
           sumF += Fmin * fraction;
         }
-      }
-      if (highit < X.end() - 1) {
-        const double upper_bin = *highit;
-        const double next_bin = *(highit + 1);
-        double fraction = (upperLimit - upper_bin);
-        if (!is_distrib) {
-          fraction /= (next_bin - upper_bin);
+      } else {
+        if (distmin > 0) {
+          const double lower_bin = *lowit;
+          const double prev_bin = *(lowit - 1);
+          double fraction = (lower_bin - lowerLimit);
+          if (!is_distrib) {
+            fraction /= (lower_bin - prev_bin);
+          }
+          const MantidVec::size_type val_index = distmin - 1;
+          sumY += Y[val_index] * fraction;
+          const double eval = E[val_index];
+          sumE += eval * eval * fraction * fraction;
+          if (rebinned_input) {
+            sumF += Fmin * fraction;
+          }
```

Comment:
Can we chunk this method and the other one into private methods. The `exec` of this file is starting to weigh in at 200+ LOC with the various nested if-else blocks

Reply:
Just pulling out this ~10 lines into a private method will involve passing these 7 variables as parameters: lowit, lowerLimit, is_distrib, distmin, Y[], E[], Fmin and also returning 3 variables sumY, sumE and sumF. This feels like a slightly ugly bit to pull out.
Having had a look at the exec, perhaps the content of the parallel for loop could be pulled as some kind of IntegrateSpectrum method?

Reply:
Yeah that would help significantly too
This isn't a PR blocker, so if it's way too much work I can be swayed against

Reply:
I've just pushed a commit with an IntegrateSpectrum method. See what you think. I can always revert to the previous commit if this is getting too big a change but seems OK?